src/thread.rs

```diff
@@ -355,14 +357,45 @@ impl<'mir, 'tcx: 'mir> ThreadManager<'mir, 'tcx> {
         // Mark the joined thread as being joined so that we detect if other
         // threads try to join it.
         self.threads[joined_thread_id].join_status = ThreadJoinStatus::Joined;
+
+        self.wait_on_thread(joined_thread_id, None, data_race)
+    }
+
+    fn wait_on_thread(
+        &mut self,
+        joined_thread_id: ThreadId,
+        timeout: Option<Time>,
+        data_race: Option<&mut data_race::GlobalState>,
+    ) -> InterpResult<'tcx> {
         if self.threads[joined_thread_id].state != ThreadState::Terminated {
             // The joined thread is still running, we need to wait for it.
             self.active_thread_mut().state = ThreadState::BlockedOnJoin(joined_thread_id);
-            trace!(
-                "{:?} blocked on {:?} when trying to join",
-                self.active_thread,
-                joined_thread_id
-            );
+
+            if let Some(timeout) = timeout {
+                self.register_timeout_callback(
+                    joined_thread_id,
+                    timeout,
+                    Box::new(move |ecx| {
+                        let state = &mut ecx.machine.threads.threads[joined_thread_id].state;
+                        assert_eq!(*state, ThreadState::BlockedOnJoin(joined_thread_id));
+                        *state = ThreadState::Enabled;
```

Comment:
I hope there's no code anywhere that assumes that `BlockedOnJoin` only completes when the thread actually joins...

Reply:
I checked and there isn't. But it doesn't matter now that timeouts are not supported anymore.